### ISSUE TITLE
Issues with certain special characters (tab and space) in deencapsulation.

### ISSUE THREAD

**Issue by TheElementalOfDestruction** [seamustuohy/RTFDE#2]
Been working on using this module in mine and found some cases where it seems to have issues. The first I noticed was then it would run across tab characters in the RTF that was supposed to go to the HTML (specifically they are written as `\'09`). While not a huge deal in terms of the HTML being functional, it does kill some of the formatting that was intended to be in the file.

The second one is a bit more complicated but is related to a translation failure, although it's unclear if the failure is with the generator or with this module. Specifically, using some instances where I have the original HTML and the encapsulated HTML, I found that `&nbsp;` is removed from the HTML stream and replaced with a literal space in the RTF stream. Here is an example of the source HTML and how it translates into RTF:
```html
<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
```
And the RTF (real example of all the characters that it translated into):
```rtf
\htmlrtf0{\*\htmltag0 <p class="MsoNormal"><o:p>}\pard\plain\htmlrtf{\f2\lang1033\fs22\htmlrtf0  {\*\htmltag0 </o:p></p>}\htmlrtf
```
Very close to the end I specifically noticed that it has a double space. Normally the space there would be ignored as it is a delimiter to the `\htmlrtf0` tag, but because it has a second one there is a literal space inserted there. In RTF, inserting a regular space in a random location that is not surrounded by text will cause an actual space to be there, however in HTML it gets ignored. Unfortunately, this has the side effect of completely breaking the spacing that the HTML document is supposed to have.

**Comment by seamustuohy** [comment]
Thanks for reporting this. Do you happen to have examples of .msg files along with the correctly formatted HTML which I can use to diagnose where the issues occur and build out unit tests as I look to fix the issue?

**Comment by TheElementalOfDestruction** [comment]
Yeah, let me see if I can find or generate some files that can reasonably be used for testing.

**Comment by TheElementalOfDestruction** [comment]
Okay, still working on ones for the second example, but I managed to generate one where the tabs are not translating correctly. Unfortunately they are actually `\tab` instead of `\'09` but it appears they are either being ignored or are being converted to spaces (or more specifically, a single space).

Here is a copy of the MSG file I made, although I made a few small modifications to strip personal data. These modifications have not affected the RTF or HTML, just the header (stripping IP addresses and such).
[Example Email #1.zip](https://github.com/seamustuohy/RTFDE/files/7967995/Example.Email.1.zip)

Also, I noticed the deencapsulated HTML tends to always have a newline character at the beginning of it. Not sure if this is supposed to happen or is some kind of bug.

**Comment by seamustuohy** [comment]
Thanks! Can you share the original (or correctly client parsed) HTML src of the email so I can use it as the canonical text the parsed data needs to match?

**Comment by TheElementalOfDestruction** [comment]
Outlook did something weird which is that even though I selected HTML generation it only generated the encapsulated form of it, so I couldn't directly access it from the file. However, I managed to get outlook to give it to me properly parsed, so here you go:

[Example Email #1 Correct Html.zip](https://github.com/seamustuohy/RTFDE/files/7974148/Example.Email.1.Correct.Html.zip)


I'm going to see if I can manage to create something that outlook will parse correctly that it didn't actually generate, as I managed to confirm that the double space does get correctly parsed into being one for a delimiter to the previous tag and one for converting to `&nbsp;`. Unfortunately the emails that have it are confidential so I'll see what I can do about getting you tests.

**Comment by TheElementalOfDestruction** [comment]
Small update: this appears to have gotten a lot more interesting. It appears the the `&nbsp;` thing is a generation error, related to some of the microsoft generators (possibly an old version that has now been patched). They incorrectly generated the encapsulation leaving the double space when it should have been producing the tag for `&nbsp;`. This is clearly visible when you look at a file that has both HTML body and RTF body. 

What makes it clear this is a generation bug is that outlook handles this the same way the module does: it sees the space and reads it literally. Took me a bit of testing to prove this, but I noticed multiple files that had this issue the generator marked in the rtf is specifically "Microsoft Exchange Server". 

New emails that I send all end up with the correct encapsulation, so not sure what can really be done about it.

**Comment by seamustuohy** [comment]
Yeah, MS believed it was OK to mess with whitespace. So, I'm not surprised. (see quote below.) Are there elements of this issue which still need to be addressed or, in your assessment, is the error unfixable on the de-encapsulation side? 

> Encapsulation and de-encapsulation can introduce changes in the content of the original document,
> as long as such changes do not affect the rendering of the document in its original format. For
> example, it is allowable to introduce, remove, or change insignificant whitespace in HTML and/or to
> normalize text line endings to use carriage return/line feed pairs (CRLFs).
- [[MS-OXRTFEX] Specification](https://interoperability.blob.core.windows.net/files/MS-OXRTFEX/%5bMS-OXRTFEX%5d.pdf)

**Comment by TheElementalOfDestruction** [comment]
Well, the screwed up because it did affect the document. Not having the white space there means that there are now paragraphs that are all squished together instead of being spaced apart nicely. 

Anyways, let me just get one of the files with `\'09` that doesn't properly keep the tabs so you can use it for testing.

**Comment by seamustuohy** [comment]
Addressed in the 0.1.0 version which is in the final notification and testing stage on [dev](https://github.com/seamustuohy/RTFDE/commit/1b29704a47e6c3a118ebb3ed2ac96e640312f8fb)
